### ISSUE TITLE
[Proposal] [Discussion] Theme builder should generate minima 1.0.1 files by default

### ISSUE THREAD

**Issue by ashmaroli** [jekyll/jekyll#5311]
Hi, 
As absurd as it sounds, I think it would be a lot more helpful to the front-end-novice when he is provided with a starting point. And that starting point is our _minima_.
Currently, only layout files (barely-there-barebones) are provided. _includes and _sass dirs are simply empty.

While he could copy those files from his gems_folder, its a lot easier to run `jekyll new-theme xyz`

files under _layouts and _includes are essentially static, hence could remain the same.  What can be treated as dynamic are the contents of _sass dir.
It contains a minima.scss and a minima dir with three partials.

In theory, Jekyll should be able to easily rename them using value from `theme.name` to **/_sass/theme_name.scss** and **/_sass/theme_name/[files]** at the same time rename **@ import** instances programmatically.

**Comment by parkr** [comment]
This defeats the purpose of gem-based themes, so I disagree that this should be the case by default.

I was thinking about proposing a command like `jekyll copy-theme-file _includes/head.html` that copies the contents of the theme's `_includes/head.html` into your site's `_includes/head.html` so you can begin to modify it with just one command. Would that satisfy your needs?

**Comment by spudowiar** [comment]
@parkr I think @ashmaroli meant that when you create a theme it should be pre-populated with the files for [minima](https://github.com/jekyll/minima)

**Comment by parkr** [comment]
@spudowiar You're right! My bad. I agree that a better starting point for `jekyll new-theme` is a great idea. Want to put together a pull request of the pieces you think we need and we can talk them over? I like the idea of starter sass, though I'm not sure about the includes. 😄

**Comment by jaybe-jekyll** [comment]
Basic {% include %}' such as required HTML, snippets, structure, etc. are valuable, sensible, and introduces the user to key (while advanced) features of the engine and ways to be more-organized and DRY.

**Comment by ashmaroli** [comment]
> I think @ashmaroli meant that when you create a theme using jekyll new-theme it should be pre-populated with the files for minima

@spudowiar, Thanks for simplifying the case into a single sentence :smiley:

**Comment by jaybe-jekyll** [comment]
https://github.com/jekyll/jekyll/pull/5317#issuecomment-244446015

Scenario:

I want to quickly spin up a new, default jekyll theme (which is minima) as my own, to start hacking on right away, locally; I don't want a gem-based theme at this moment-- I want it local and in my new jekyll source.

I'd like to, i.e.:

jekyll new --local mylocaltheme
I'd like that to produce a new directory, ./mylocaltheme, containing jekyll new defaults (sans theme: minima in _config.yml) along with local copies of minima gem and related directories/files (includes, etc.) within ./mylocaltheme.

I cd ./mylocaltheme and begin configuring/changing my new theme source starting point.

**Comment by jaybe-jekyll** [comment]
Would that then also imply an additional `--theme` switch?

i.e. `jekyll new --local --theme [minima|anothergem] mylocaltheme` ?

**Comment by pathawks** [comment]
> I want to quickly spin up a new, default jekyll theme (which is minima) as my own, to start hacking on right away, locally; I don't want a gem-based theme at this moment-- I want it local and in my new jekyll source.

Can you explain this in a bit more detail? It sounds like what you are asking for is to create a new _site_ that does not depend on an external theme. I’m trying to understand how that relates to creating a new _theme_.

Please excuse me if I am misunderstanding.

**Comment by jaybe-jekyll** [comment]
@pathawks 

Correct.

I would like to begin creating my great new theme.

I'd like to start with the default minima theme as my base.

To do that I need a `jekyll new` site/framework, *without* `theme: minima` in `_config.yml`, and the jekyll components that make up the minima theme. i.e. `_layouts`, `_includes`, etc.

Everything, including the "theme", will be served from `./mynewproject`.

I'd get the current/latest minima theme, pulled/copied from the minima gem, and included in `jekyll new mynewproject`, via, perhaps: `jekyll new --local --theme minima mynewproject`.

The counter thinking/argument would be to expect [new] users to understand how/where gems work and are stored. In my case, sometimes, ~/.gems/.../.../.../bin/...

Does this help; can I further clarify?

**Comment by pathawks** [comment]
> Does this help; can I further clarify?

Ok, I understand. I thought that this discussion was about the workflow for creating a new theme, but it seems I was mistaken.

Carry on 😁:+1:

**Comment by jaybe-jekyll** [comment]
It technically could/would/eventually be about that as well!

https://github.com/jekyll/jekyll/pull/5317#issuecomment-244450584

On deck/heads up/next:

> I created an awesome theme with this new command and functionality, now I want to create/bundle/package my theme into a gem so it can be a shared and installable jekyll gem-based theme! How do I do that?

**Comment by DirtyF** [comment]
@jaybe-jekyll I like the idea of a kinda `jekyll clone [theme] [path]` to get started from a boilerplate

**Comment by ashmaroli** [comment]
> I thought that this discussion was about the workflow for creating a new theme

@pathawks, When I opened this ticket, my intention was indeed to have [ the files I would copy from minima gem ] already loaded inside the new_theme directory. I made the PR to that effect.
And you correctly expressed your opinion based on my approach.

---
@jaybe-jekyll, I like the idea of starting off with a new blog configured to use a theme under development. It makes sense. 
( I `jekyll new-theme xyz` into the blog directory, so never thought about this approach. :+1:  )
[ Note: Jekyll already looks into root for [theme_related_files], and uses them regardless of whether a theme is configured. ] 
w/o changing much, instead of `jekyll new --local mylocaltheme`, how about a more semantic 
switch `--new-theme` ? 
```
jekyll new my_blog --new-theme
```
I suggest opening a new ticket to discuss this. :smile: 

---
@DirtyF :+1:  for suggesting those options.

**Comment by spudowiar** [comment]
I now have a patch that allows you to do the following two things

Create a theme based on Minima
```
bundle exec jekyll new-theme -b minima custom-theme
```

Create a site using a mutable theme
```
bundle exec jekyll new .
bundle exec jekyll new-theme -b minima custom-theme _theme
echo 'gem "custom-theme", path: "_theme"' >> Gemfile
sed 's/minima/custom-theme/g' -i _config.yml
```

**Comment by ashmaroli** [comment]
@spudowiar,  :+1:  at a glance. Can you open a PR for this patch? Would love to see the source code.

**Comment by spudowiar** [comment]
@ashmaroli Need to rebase into a few commits, bit busy at the moment so I thought I'd just put that up there for now

**Comment by ashmaroli** [comment]
will close this after #5319 gets merged.

**Comment by jekyllbot** [comment]
This issue has been automatically marked as stale because it has not been commented on for at least two months.

The resources of the Jekyll team are limited, and so we are asking for your help.

If this is a **bug** and you can still reproduce this error on the <code>3.3-stable</code> or <code>master</code> branch, please reply with all of the information you have about it in order to keep the issue open.

If this is a **feature request**, please consider building it first as a plugin. Jekyll 3 introduced [hooks](http://jekyllrb.com/docs/plugins/#hooks) which provide convenient access points throughout the Jekyll build pipeline whereby most needs can be fulfilled. If this is something that cannot be built as a plugin, then please provide more information about why in order to keep this issue open.

This issue will automatically be closed in two months if no further activity occurs. Thank you for all your contributions.
